chain/client/src/client.rs

```diff
@@ -139,9 +151,12 @@ pub struct Client {
     /// Approvals for which we do not have the block yet
     pub pending_approvals:
         lru::LruCache<ApprovalInner, HashMap<AccountId, (Approval, ApprovalType)>>,
+    /// A mapping from an epoch that we know needs to be state synced for some shards
+    /// to a tracker that will find an appropriate sync_hash for state sync to that epoch
+    catchup_tracker: HashMap<EpochId, NewChunkTracker>,
```

Comment:
You know, that is a good question... It seems this mapping from epoch ID to catchup status has been around since state sync was first implemented, and looking at the code it doesn't seem to me like it should ever have more than one... When we add one, it's because we see a block in a new epoch T, and we know we're going to be a chunk producer for T+1 in a new shard. The epoch ID for T+1  is the hash of the last block in epoch T-1, so if there's a fork at the beginning of the epoch, it will still have the same epoch ID. So that should mean that for any given epoch height, we only have one of these. And then if we ask if it's possible to have epoch T and T+1 in there at the same time, we can look at the fact that we remove the epoch info for epoch T in `finish_catchup_blocks()` when we call `remove_state_sync_info()`. If that hasn't happened by the time we get to the first block of epoch T+1, we will not add another catchup state keyed by the epoch ID of epoch T+1, because we'll [call that block an orphan](https://github.com/near/nearcore/blob/bf28607aa18b59e4441b61e9ffb73b4cdc18c1fc/chain/chain/src/chain.rs#L2344) until [we remove the last block of epoch T](https://github.com/near/nearcore/blob/bf28607aa18b59e4441b61e9ffb73b4cdc18c1fc/chain/chain/src/chain.rs#L3284) from the `BlocksToCatchup` column in `finish_catchup_blocks()`  .

So idk how it is even possible to have two, but maybe I'm missing something and that logic was put there for a reason?

But now that I think about it, what happens if we apply the first block of a new epoch and then save `StateSyncInfo` for it, which on the DB is keyed by the hash of that first block, and then that block doesn't end up on the main chain, because there's a fork off the last block of epoch T-1? Is there some reason that won't happen for the first block of an epoch? I'm not sure if there is, but if so, then there's an implicit assumption here that at least should have gotten a comment explaining, and if not, there's a potential bug here. I think in general maybe we should only be working with final blocks when doing all this state sync stuff. If it is a bug, then for now it's a pretty unlikely one I guess, but it's worth investigating more. Shouldn't be too too hard to try causing a fork at the first block of an epoch in a testloop test and seeing what happens

Reply:
ehh it's possible that last part is not a bug actually, since there's care put into only removing the particular sync hash from the `BlocksToCatchup` [here](https://github.com/near/nearcore/blob/bf28607aa18b59e4441b61e9ffb73b4cdc18c1fc/chain/chain/src/chain.rs#L3275), but it's a bit confusing and worth sanity checking